File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));


    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());
        int[] nums = new int[n];
        Stack<Integer> st = new Stack<>();
        List<Character> answers = new ArrayList<>();

        for (int i = 0; i < n; ++i) {
            nums[i] = Integer.parseInt(br.readLine());
        }

        int curr = 0;
    
        for (int i = 1; i <= n; ++i) {
            st.add(i);
            answers.add('+');
            while(!st.isEmpty() && st.peek() == nums[curr]) {
                answers.add('-');
                st.pop();
                ++curr;
            }
        }
        
        if (st.isEmpty()) {
            for (char ch : answers) {
                bw.write(ch + "\n");
            }
        }
        else {
            bw.write("NO\n");
        }

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.`out`))
val st: Stack<Int> = Stack()
val data = ArrayList<Int>()
val answer = ArrayList<Char>()
fun main() {

    val n = br.readLine().toInt()
    var idx = 0

    for (i in 1..n) {
        data.add(br.readLine().toInt())
    }

    for (i in 1..n) {
        st.push(i)
        answer.add('+')

        while(st.isNotEmpty() && st.peek() == data[idx]) {
            st.pop()
            answer.add('-')
            ++idx
        }
    }

    if (st.isEmpty()) {
        for (ch in answer) {
            bw.write("$ch\n")
        }
    } else {
        bw.write("NO\n")
    }

    bw.flush()
    bw.close()
    br.close()
}


